# Project: To predict the sales of weather sensitive products of walmart 
------------------------------------------------------------------------------------------------------------------

## PHASE 2: EDA - weather.csv
------------------------------------------------------------------------------------------------------------------

### Author: Soumen Chatterjee
==================================================================================================================


#### The main goal here is - 

1. Import all the req libraries
2. Review the weather.csv data file
3. Clean the dataframe by removing null values, converting the data types to appropriate one
4. Feature engineering - feature encoding
5. Filling up Null values with center secific Median values
6. Dropping redundant columns
7. Saving the cleaned weather file

### Importing all the req. libraries

In [ ]:
import pandas as pd
import numpy as np
from datetime import datetime
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import sklearn
from pandas.io.formats.format import Datetime64Formatter
import datetime as dt
import warnings
warnings.filterwarnings("ignore")


### Reviewing "weather.csv":

In [ ]:
#Loading the data
from google.colab import drive
drive.mount('/content/drive')

df_weather = pd.read_csv("/content/drive/MyDrive/Project_TS_Walmart/Phase 2/Data/weather.csv")

#Checking how it appears
df_weather.head(5)

Mounted at /content/drive


station_nbr        date tmax tmin  ... sealevel resultspeed resultdir avgspeed
0            1  2012-01-01   52   31  ...    29.92         3.6        20      4.6
1            2  2012-01-01   48   33  ...    29.91         9.1        23     11.3
2            3  2012-01-01   55   34  ...    30.47         9.9        31     10.0
3            4  2012-01-01   63   47  ...    30.48         8.0        35      8.2
4            6  2012-01-01   63   34  ...    30.47        14.0        36     13.8

[5 rows x 20 columns]

#Checking the base data structure

In [ ]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20517 entries, 0 to 20516
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   station_nbr  20517 non-null  int64 
 1   date         20517 non-null  object
 2   tmax         20517 non-null  object
 3   tmin         20517 non-null  object
 4   tavg         20517 non-null  object
 5   depart       20517 non-null  object
 6   dewpoint     20517 non-null  object
 7   wetbulb      20517 non-null  object
 8   heat         20517 non-null  object
 9   cool         20517 non-null  object
 10  sunrise      20517 non-null  object
 11  sunset       20517 non-null  object
 12  codesum      20517 non-null  object
 13  snowfall     20517 non-null  object
 14  preciptotal  20517 non-null  object
 15  stnpressure  20517 non-null  object
 16  sealevel     20517 non-null  object
 17  resultspeed  20517 non-null  object
 18  resultdir    20517 non-null  object
 19  avgspeed     20517 non-nu

Findings:
1. If we review the dataframe we can see that - apart from 'date' & 'codesum' coulmn rest all should be treated as int64, but they are being treated as string

2. It is because the missing values are updated as 'M',"-", " "

3. We will replace 'M',"-", " " from throughout the data frame, then will convert all the columns to int64 for further proceedings.

In [ ]:
df_weather.replace(to_replace ="M", value = np.nan,inplace=True)
df_weather.replace(to_replace ="  T", value = np.nan,inplace=True)
df_weather.replace('-', np.nan,inplace=True)
df_weather.replace(' ', np.nan,inplace=True)
#df_weather.iloc[412]
df_weather.info() #As we have removed the unwanted charecters which filled the Null places, so we can see the Null rows now

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20517 entries, 0 to 20516
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   station_nbr  20517 non-null  int64 
 1   date         20517 non-null  object
 2   tmax         19611 non-null  object
 3   tmin         19609 non-null  object
 4   tavg         19048 non-null  object
 5   depart       9006 non-null   object
 6   dewpoint     19851 non-null  object
 7   wetbulb      19265 non-null  object
 8   heat         19048 non-null  object
 9   cool         19048 non-null  object
 10  sunrise      10861 non-null  object
 11  sunset       10861 non-null  object
 12  codesum      9466 non-null   object
 13  snowfall     12982 non-null  object
 14  preciptotal  17247 non-null  object
 15  stnpressure  19588 non-null  object
 16  sealevel     18793 non-null  object
 17  resultspeed  19928 non-null  object
 18  resultdir    19928 non-null  object
 19  avgspeed     19642 non-nu

We will convert the data type of string features to numeric now

In [ ]:
# Another way of converting one by one - df_weather["tmax"] = df_weather['tmax'].astype(float)
# using dictionary to convert specific columns
convert_dict = {'tmax': float,'tmin': float,'tavg': float,'depart': float,
                'dewpoint': float,'wetbulb': float,'heat': float,'cool': float,
                'snowfall': float,'preciptotal': float,'stnpressure': float,'sealevel': float,
                'resultspeed': float,'resultdir': float,'avgspeed': float
               }
df_weather = df_weather.astype(convert_dict) 

Tried converting the Sunrise & Sunset columns to dateframe getting failed. As we have only 10861 rows fill, that is half of the rows are filled only, so we can drop these 2 columns

In [ ]:
#df_weather["date"] = pd.to_datetime(df_weather["date"])
#df_weather["sunrise"] =(pd.to_datetime(df_weather['sunrise'],format='%H%H%M%M'))
#a=(pd.to_datetime(df_weather['sunrise'],format='%H%M'))
#a=datetime.strptime(df_weather['sunrise'],'%H%M')
#df = pd.DataFrame(df_weather, columns = list('sunrise'))
#df.apply(datetime.strptime, args = [1])
#datetime.strptime(df_weather['sunrise'],'%H%M')
#pd.to_datetime(df,format='%H%M')
#df_weather['sunset'] = pd.to_datetime(df_weather['sunset'], format='%H%M' ).dt.time
#df_weather["sunrise"] = df_weather['sunrise'].str.strip()
#dummy = df_weather['sunrise'].dropna()
#pd.to_datetime(dummy, format='%H%M' )
#import datetime as dt
#df_weather.sunrise = df_weather.sunrise.apply(lambda x: '2016-01-01 ' + str(x))
#df_weather.sunrise = pd.to_datetime(df_weather.sunrise).apply(lambda x: dt.datetime.time(x))


"codesum" feature is having multi option categorical feature, we will need to do one hot encoding.To do that - 

1. We first finding out possible number of Unique values - 29 in this case
2. Then will create extra columns for one hot encoding

In [ ]:
lstCodesum = list(df_weather.codesum.unique())
#lstCodesum
codesum_set = set()
for codes in lstCodesum:
  lst = str(codes).split(" ") #list (str((codes.split(" "))))
  #print(codes)
  for ele in lst:
    codesum_set.add(ele)

codesum_list = list(codesum_set)
len(codesum_set)
#Total number of Unique values in codesum column

29

Complist: This holds exsisting and newly added columns to have one hot encoding

In [ ]:
y = pd.get_dummies(codesum_list, prefix='codesum')
y.columns

Complist = list(df_weather.columns) + list(y.columns)


Creating a fresh dataframe with revised list of columns that is Complist & doing one hot encoding

Removing base codesum column (string)

In [ ]:
df_weather2 = df_weather.reindex(columns = Complist, copy=True)
#df_weather2 = df_weather2.iloc[:5]
rw = 0
for index, row in df_weather2.iterrows(): 
    chklist = str(row['codesum']).split(" ")
    for cde in codesum_list:
      if cde in chklist:
        col = "codesum_"+cde
        df_weather2.iat[rw, df_weather2.columns.get_loc(col)] = 1
    rw += 1
      

# Removing column name 'codesum'
df_weather2.drop(['codesum'], axis = 1, inplace=True)
df_weather2.head(5) 

station_nbr        date  tmax  ...  codesum_VCFG  codesum_VCTS  codesum_nan
0            1  2012-01-01  52.0  ...           NaN           NaN          NaN
1            2  2012-01-01  48.0  ...           NaN           NaN          NaN
2            3  2012-01-01  55.0  ...           NaN           NaN          1.0
3            4  2012-01-01  63.0  ...           NaN           NaN          1.0
4            6  2012-01-01  63.0  ...           NaN           NaN          1.0

[5 rows x 48 columns]

Again tried changing the Sunset & Sunrise columns to datetime format

In [ ]:
df_weather3 = df_weather2.copy()
'''
maxlen = len(df_weather3['sunset'])
for row in range(0,maxlen):
  try:
    timeDF=(pd.to_datetime(df_weather3['sunrise'][row], format='%H%M'))
    df_weather3['sunrise'][row] = pd.to_datetime(timeDF)#dt.datetime.time(timeDF)

    timeDF2=(pd.to_datetime(df_weather3['sunset'][row], format='%H%M'))
    df_weather3['sunset'][row] = pd.to_datetime(timeDF2)#dt.datetime.time(timeDF)
  except:
    df_weather3['sunrise'][row] = np.datetime64('NaT')
    df_weather3['sunset'][row] = np.datetime64('NaT')
'''
print("Not used")


Not used


In [ ]:
#abc = (pd.to_datetime(df_weather3['sunrise'], format='%Y-%m-%d %H:%M:%S'))#df_weather3['sunrise']



In [ ]:
#df_weather3.info()

Creating station wise Median Dataframe, which will be used to fillna

In [ ]:
df_weather_median = df_weather3.groupby(['station_nbr']).median()
df_weather_median

tmax  tmin  tavg  ...  codesum_VCFG  codesum_VCTS  codesum_nan
station_nbr                    ...                                         
1            63.0  43.0  54.0  ...           NaN           1.0          1.0
2            61.0  43.0  53.0  ...           NaN           1.0          1.0
3            76.0  54.0  66.0  ...           1.0           1.0          1.0
4            85.0  63.0  74.0  ...           NaN           1.0          1.0
5            70.0  51.0  60.0  ...           NaN           NaN          1.0
6            84.0  60.0  72.0  ...           NaN           1.0          1.0
7            77.0  55.0  67.0  ...           NaN           1.0          1.0
8            84.0  64.0  77.0  ...           NaN           1.0          1.0
9            60.0  40.0  51.0  ...           1.0           1.0          1.0
10           83.0  67.0  75.0  ...           NaN           1.0          1.0
11           82.0  61.0  72.0  ...           NaN           1.0          1.0
12           83.0  63.5  74.0  ...           NaN           1.0          1.0
13           67.0  38.0  53.0  ...           NaN           1.0          1.0
14           77.0  54.0  66.0  ...           NaN           1.0          1.0
15           62.0  37.0  50.0  ...           NaN           1.0          1.0
16           62.0  42.0  52.0  ...           NaN           1.0          1.0
17           77.0  53.0  66.0  ...           NaN           1.0          1.0
18           82.0  62.0  72.0  ...           1.0           1.0          1.0
19           59.0  42.0  51.0  ...           NaN           NaN          1.0
20           82.0  60.0  71.5  ...           NaN           1.0          1.0

[20 rows x 44 columns]

There are few station where we have NaN in Median dataframe.
Now using the first Median variable that is df_weather_median and filling NaN with Median of df_weather_median

In [ ]:
df_weather_median2 = df_weather_median.copy()
df_weather_median2.loc[df_weather_median2['depart'].isnull(),'depart'] = df_weather_median2.loc[df_weather_median2['depart'].isnull(),'depart'].fillna(df_weather_median.depart.median())
df_weather_median2.loc[df_weather_median2['wetbulb'].isnull(),'wetbulb'] = df_weather_median2.loc[df_weather_median2['wetbulb'].isnull(),'wetbulb'].fillna(df_weather_median.wetbulb.median())
df_weather_median2.loc[df_weather_median2['snowfall'].isnull(),'snowfall'] = df_weather_median2.loc[df_weather_median2['snowfall'].isnull(),'snowfall'].fillna(df_weather_median.snowfall.median())

df_weather_median2.loc[df_weather_median2['stnpressure'].isnull(),'stnpressure'] = df_weather_median2.loc[df_weather_median2['stnpressure'].isnull(),'stnpressure'].fillna(df_weather_median.stnpressure.median())
df_weather_median2.loc[df_weather_median2['sealevel'].isnull(),'sealevel'] = df_weather_median2.loc[df_weather_median2['sealevel'].isnull(),'sealevel'].fillna(df_weather_median.sealevel.median())
df_weather_median2

tmax  tmin  tavg  ...  codesum_VCFG  codesum_VCTS  codesum_nan
station_nbr                    ...                                         
1            63.0  43.0  54.0  ...           NaN           1.0          1.0
2            61.0  43.0  53.0  ...           NaN           1.0          1.0
3            76.0  54.0  66.0  ...           1.0           1.0          1.0
4            85.0  63.0  74.0  ...           NaN           1.0          1.0
5            70.0  51.0  60.0  ...           NaN           NaN          1.0
6            84.0  60.0  72.0  ...           NaN           1.0          1.0
7            77.0  55.0  67.0  ...           NaN           1.0          1.0
8            84.0  64.0  77.0  ...           NaN           1.0          1.0
9            60.0  40.0  51.0  ...           1.0           1.0          1.0
10           83.0  67.0  75.0  ...           NaN           1.0          1.0
11           82.0  61.0  72.0  ...           NaN           1.0          1.0
12           83.0  63.5  74.0  ...           NaN           1.0          1.0
13           67.0  38.0  53.0  ...           NaN           1.0          1.0
14           77.0  54.0  66.0  ...           NaN           1.0          1.0
15           62.0  37.0  50.0  ...           NaN           1.0          1.0
16           62.0  42.0  52.0  ...           NaN           1.0          1.0
17           77.0  53.0  66.0  ...           NaN           1.0          1.0
18           82.0  62.0  72.0  ...           1.0           1.0          1.0
19           59.0  42.0  51.0  ...           NaN           NaN          1.0
20           82.0  60.0  71.5  ...           NaN           1.0          1.0

[20 rows x 44 columns]

Filling up missing values with Station specific Median values using df_weather_median2

In [ ]:
df_weather4 = df_weather3.copy()

columnList = ["tmax", "tmin", "tavg", "depart", "dewpoint", "wetbulb", "heat", "cool", "snowfall", "preciptotal", "stnpressure", "sealevel", "resultspeed", "resultdir", "avgspeed"]
for station_nbr in range(1,21):
  for clmn in columnList:
    condition = (df_weather4[clmn].isnull()) & (df_weather4['station_nbr'] == station_nbr)
    df_weather4.loc[condition,clmn] = df_weather4.loc[condition,clmn].fillna(df_weather_median2.iloc[station_nbr-1][clmn])




Filling up One Hot encoding null values with 0. So that the columns only contains boolean values

In [ ]:
df_weather4.update(df_weather3[['codesum_BCFG',	'codesum_BLDU',	'codesum_BLSN',	'codesum_BR',	'codesum_DU',	'codesum_DZ',	'codesum_FG',	'codesum_FG+',	'codesum_FU',	'codesum_FZDZ',	'codesum_FZFG',	'codesum_FZRA',	'codesum_GR',	'codesum_GS',	'codesum_HZ',	'codesum_MIFG',	'codesum_PL',	'codesum_PRFG',	'codesum_RA',	'codesum_SG',	'codesum_SN',	'codesum_SQ',	'codesum_TS',	'codesum_TSRA',	'codesum_TSSN',	'codesum_UP',	'codesum_VCFG',	'codesum_VCTS',	'codesum_nan']].fillna(0))
df_weather4

station_nbr        date  tmax  ...  codesum_VCFG  codesum_VCTS  codesum_nan
0                1  2012-01-01  52.0  ...           0.0           0.0          0.0
1                2  2012-01-01  48.0  ...           0.0           0.0          0.0
2                3  2012-01-01  55.0  ...           0.0           0.0          1.0
3                4  2012-01-01  63.0  ...           0.0           0.0          1.0
4                6  2012-01-01  63.0  ...           0.0           0.0          1.0
...            ...         ...   ...  ...           ...           ...          ...
20512           16  2014-10-31  53.0  ...           0.0           0.0          1.0
20513           17  2014-10-31  59.0  ...           0.0           0.0          0.0
20514           18  2014-10-31  67.0  ...           0.0           0.0          1.0
20515           19  2014-10-31  45.0  ...           0.0           0.0          0.0
20516           20  2014-10-31  68.0  ...           0.0           0.0          1.0

[20517 rows x 48 columns]

Droping Sunset & Sunrise columns

In [ ]:

df_weather5 = df_weather4.copy()
df_weather5.drop(['sunrise','sunset'], axis = 1,inplace=True)


Now the weather dataframe is cleaned

Creating Different features - Day, Month, Quarter, Holiday

In [ ]:
df_weather6 = df_weather5.copy()
df_weather6["date"] = pd.to_datetime(df_weather6["date"])
df_weather6["Day"] = df_weather6.date.dt.day
df_weather6["Month"] = df_weather6.date.dt.month
df_weather6["Quarter"] = df_weather6.date.dt.quarter

from pandas.tseries.holiday import USFederalHolidayCalendar as calendar

dr = pd.date_range(start='2012-01-01', end='2014-10-31')
cal = calendar()
holidays = cal.holidays(start=dr.min(), end=dr.max())

df_weather6['Holiday'] = df_weather6['date'].isin(holidays)

In [ ]:
df_weather6.tail(5)
df_weather6.columns

Index(['station_nbr', 'date', 'tmax', 'tmin', 'tavg', 'depart', 'dewpoint',
       'wetbulb', 'heat', 'cool', 'snowfall', 'preciptotal', 'stnpressure',
       'sealevel', 'resultspeed', 'resultdir', 'avgspeed', 'codesum_BCFG',
       'codesum_BLDU', 'codesum_BLSN', 'codesum_BR', 'codesum_DU',
       'codesum_DZ', 'codesum_FG', 'codesum_FG+', 'codesum_FU', 'codesum_FZDZ',
       'codesum_FZFG', 'codesum_FZRA', 'codesum_GR', 'codesum_GS',
       'codesum_HZ', 'codesum_MIFG', 'codesum_PL', 'codesum_PRFG',
       'codesum_RA', 'codesum_SG', 'codesum_SN', 'codesum_SQ', 'codesum_TS',
       'codesum_TSRA', 'codesum_TSSN', 'codesum_UP', 'codesum_VCFG',
       'codesum_VCTS', 'codesum_nan', 'Day', 'Month', 'Quarter', 'Holiday'],
      dtype='object')

In [ ]:
# saving the dataframe
df_weather6.to_csv(r'/content/drive/MyDrive/Project_TS_Walmart/Phase 2/Data/weather_cleaned.csv', index=False)

In [ ]:
df_weather6

station_nbr       date  tmax  tmin  ...  Day  Month  Quarter  Holiday
0                1 2012-01-01  52.0  31.0  ...    1      1        1    False
1                2 2012-01-01  48.0  33.0  ...    1      1        1    False
2                3 2012-01-01  55.0  34.0  ...    1      1        1    False
3                4 2012-01-01  63.0  47.0  ...    1      1        1    False
4                6 2012-01-01  63.0  34.0  ...    1      1        1    False
...            ...        ...   ...   ...  ...  ...    ...      ...      ...
20512           16 2014-10-31  53.0  34.0  ...   31     10        4    False
20513           17 2014-10-31  59.0  34.0  ...   31     10        4    False
20514           18 2014-10-31  67.0  49.0  ...   31     10        4    False
20515           19 2014-10-31  45.0  33.0  ...   31     10        4    False
20516           20 2014-10-31  68.0  50.0  ...   31     10        4    False

[20517 rows x 50 columns]

In [ ]:
df_weather6[['tmax', 'tmin', 'tavg', 'depart', 'dewpoint','wetbulb', 'heat', 'cool', 'snowfall', 'preciptotal', 'stnpressure','sealevel', 'resultspeed', 'resultdir', 'avgspeed']].describe()

tmax          tmin  ...     resultdir      avgspeed
count  20517.000000  20517.000000  ...  20517.000000  20517.000000
mean      71.570649     50.679168  ...     18.818663      7.983869
std       19.415409     18.719377  ...      9.558794      3.865889
min      -11.000000    -21.000000  ...      1.000000      0.000000
25%       60.000000     37.000000  ...     13.000000      5.200000
50%       75.000000     52.000000  ...     18.000000      7.500000
75%       86.000000     66.000000  ...     26.000000      9.900000
max      114.000000     88.000000  ...     36.000000     28.700000

[8 rows x 15 columns]